arch: arm: cortex-m: add support for clearing NXP MPU regions at boot

henrikbrixandersen opened the pull request:
Clear NXP MPU regions at boot if CONFIG_INIT_ARCH_HW_AT_BOOT is enabled. Also remove MPU support from the K2x SoC since this leads to compilation issues as the K22F does not contain an MPU.

Fixes: #34045

henrikbrixandersen committed:
arch: arm: cortex-m: add support for clearing NXP MPU regions at boot

Clear NXP MPU regions at boot if CONFIG_INIT_ARCH_HW_AT_BOOT is
enabled.

Fixes: #34045

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
This brings https://github.com/mcu-tools/mcuboot/pull/896 into Zephyr.

nandojve commented:
Hello @henrikbrixandersen ,

First, thank you so much for sending this patch.

Unfortunately, I still need select `MCUBOOT_CLEANUP_ARM_CORE` at MCUboot.

My Steps:

```
# [ 1 ] build normally doesn't boot image

# At Zephyr dir after fetch PR
git checkout #34680
git fetch upstream
git rebase upstream/master
west update
cd ..
ls
bootloader  modules  net-tools  tools  zephyr

# build/flash bootloader
west build -b frdm_k64f -d build/mcuboot-frdm_k64f bootloader/mcuboot/boot/zephyr
west flash -d build/mcuboot-frdm_k64f

# build/flash updatehub
west build -b frdm_k64f -d build/app zephyr/samples/subsys/mgmt/updatehub
west sign -t imgtool -d build/app --no-hex -- --version 1.0.0 --pad --key bootloader/mcuboot/root-rsa-2048.pem
west flash -d build/app --bin-file build/app/zephyr/zephyr.signed.bin
```

```
# [ 2 ] build with MCUboot clean chain load enabled

rm -rf build

west build -b frdm_k64f -d build/mcuboot-frdm_k64f bootloader/mcuboot/boot/zephyr -DCONFIG_MCUBOOT_CLEANUP_ARM_CORE=y
west flash -d build/mcuboot-frdm_k64f

# build/flash updatehub
west build -b frdm_k64f -d build/app zephyr/samples/subsys/mgmt/updatehub
west sign -t imgtool -d build/app --no-hex -- --version 1.0.0 --pad --key bootloader/mcuboot/root-rsa-2048.pem
west flash -d build/app --bin-file build/app/zephyr/zephyr.signed.bin
```

The [ 2 ] boot Zephyr. However, I still see the issues. frdm-k64f not confirm image and the consequence is board boot on a loop.
```
*** Booting Zephyr OS build zephyr-v2.5.0-2840-g363af20fc4a6  ***
I: Starting bootloader
I: Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: primary slot
I: Swap type: none
I: Bootloader chainload address offset: 0x20000
I: Jumping to the first image slot
*** Booting Zephyr OS build zephyr-v2.5.0-2840-g363af20fc4a6  ***


uart:~$ updatehub info 
Unique device id: ffffffff4e454130000b000f
Firmware Version: 1.0.0
Product uid: 
UpdateHub Server: coap.updatehub.io
[00:00:03.001,000] <inf> eth_mcux: ETH_0 enabled 100M full-duplex mode.
[00:00:03.002,000] <inf> main: UpdateHub sample app started
[00:00:03.002,000] <inf> main: Confirming the boot image
[00:00:03.002,000] <err> main: Error to confirm the image
[00:00:03.002,000] <inf> main: Network disconnected
uart:~$ 
[00:00:11.006,000] <inf> main: Network connected
[00:00:11.006,000] <inf> main: Starting UpdateHub polling mode
[00:00:11.006,000] <inf> updatehub: SHA-256 verification on download and from flash
[00:00:11.006,000] <err> updatehub: The current image is not confirmed
[00:00:11.006,000] <err> updatehub: Image is unconfirmed. Rebooting to revert back to previousconfirmed image.*** Booting Zephyr OS build zephyr-v2.5.0-2840-g363af20fc4a6  ***
I: Starting bootloader
I: Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: primary slot
I: Swap type: none
I: Bootloader chainload address offset: 0x20000
I: Jumping to the first image slot
*** Booting Zephyr OS build zephyr-v2.5.0-2840-g363af20fc4a6  ***


[00:00:03.001,000] <inf> eth_mcux: ETH_0 enabled 100M full-duplex mode.
[00:00:03.002,000] <inf> main: UpdateHub sample app started
[00:00:03.002,000] <inf> main: Confirming the boot image
[00:00:03.002,000] <err> main: Error to confirm the image
[00:00:03.002,000] <inf> main: Network disconnected
uart:~$ 
```

ioannisg reviewed (changes requested):
Totally fine with the change, of course, but can you move this out of ARCH code? We try to remove SoC-specific code from core AARCH32 directory; definitely not add more :)
You could implement the function either under soc/arm/common or soc/arm/nxp_kinetis.

ioannisg commented:
> Clear NXP MPU regions at boot if CONFIG_INIT_ARCH_HW_AT_BOOT is enabled.

This call is not implemented (is not part of the PR, at least)

henrikbrixandersen committed:
soc: arm: nxp: kinetis: k2x: remove MPU support

Remove (disabled) MPU support for the NXP Kinetis K2x. At least the
MK22F51212, which is the only K2x supported by Zephyr so far, does not
contain an MPU.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
> > Clear NXP MPU regions at boot if CONFIG_INIT_ARCH_HW_AT_BOOT is enabled.
> 
> This call is not implemented (is not part of the PR, at least)

@ioannisg Not sure what you mean with the above comment?

henrikbrixandersen commented:
> Totally fine with the change, of course, but can you move this out of ARCH code? We try to remove SoC-specific code from core AARCH32 directory; definitely not add more :)
> You could implement the function either under soc/arm/common or soc/arm/nxp_kinetis.

Currently, the NXP MPU handling is all done under arch specific code (try grep'ing for `CONFIG_CPU_HAS_NXP_MPU`).
How would you suggest I tie in the NXP-specific functions to `z_arm_init_arch_hw_at_boot()`?

While I agree that SoC specific code does not belong there, I am not sure I can fit in cleaning up the NXP MPU handling in the scope of this PR.

henrikbrixandersen commented:
> > Clear NXP MPU regions at boot if CONFIG_INIT_ARCH_HW_AT_BOOT is enabled.
> 
> This call is not implemented (is not part of the PR, at least)

Fixed (for now, at least, pending a potential shuffle of where the code must reside).

nandojve commented:
Hello @henrikbrixandersen ,
I confirmed that current PR, on top of faf68fb213, now boot image.

I still have the issue [34683](https://github.com/zephyrproject-rtos/zephyr/issues/34683), which don't confirm the image, but this is another history.

Thank you so much for fix this!

CC: @otavio

ioannisg commented:
Let's treat this as a bug (it is actually, since chain-loadable images need to ensure the MPU state is clean), so we are not bound by the feature freeze.

mmahadevan108 reviewed (approved):
Thank you.

galak commented:
@henrikbrixandersen / @ioannisg what's the status here?

henrikbrixandersen commented:
> @henrikbrixandersen / @ioannisg what's the status here?

From my point of view it is ready to go in. I think splitting the NXP-specific MPU code from the arch code needs to happen in another PR.

dleach02 commented:
It feels like the z_arm_clear_arm_mpu_config() is in the wrong file to begin with and should be split out of sbc.c. 

The NXP version should be in arch/arm/core/aarch32/mpu/nxp_mpu.c
Baseline Arm version in arch/arm/core/aarch32/aarch32/mpu/arm_mpu.c

This would follow the similar model of the rest of the MPU functions.

henrikbrixandersen commented:
> It feels like the z_arm_clear_arm_mpu_config() is in the wrong file to begin with and should be split out of sbc.c.
> 
> The NXP version should be in arch/arm/core/aarch32/mpu/nxp_mpu.c
> Baseline Arm version in arch/arm/core/aarch32/aarch32/mpu/arm_mpu.c
> 
> This would follow the similar model of the rest of the MPU functions.

If this is the consensus, I can make those changes in this PR. Will this resolve your concerns, @ioannisg?

ioannisg commented:
> > It feels like the z_arm_clear_arm_mpu_config() is in the wrong file to begin with and should be split out of sbc.c.
> > The NXP version should be in arch/arm/core/aarch32/mpu/nxp_mpu.c
> > Baseline Arm version in arch/arm/core/aarch32/aarch32/mpu/arm_mpu.c
> > This would follow the similar model of the rest of the MPU functions.
> 
> If this is the consensus, I can make those changes in this PR. Will this resolve your concerns, @ioannisg?

@henrikbrixandersen will look at it after the release meeting

galak commented:
I'm in agreement with @dleach02 but will not hold up this getting merged for 2.6 to fix the issue and doing the code cleanup post release.

henrikbrixandersen commented:
@ioannisg Any update on this?

ioannisg commented:
> @ioannisg Any update on this?

Yeap, my next one for review, hang on a sec ;)

MaureenHelm reviewed (approved):
Let's get this in before rc2 so it can be included in board farm testing.

ioannisg reviewed (approved):
agreed. We can clean this up later.
I don't really like that the function name is z_arm_clear_arm_mpu_config, and is used for the nxp mpu; it should be a different function name. but no big deal, cause this needs further cleanup, anyways.
